### PR TITLE
DE458895: Adding generic entity support

### PR DESCRIPTION
**Pull request by nikhilahuja**
## Description
Generic entity support

## Checklist
- [x] Pull Request Created
- [x] Unit tests have been added
- [x] Integration/Functional test cases have been written and automated

### CODE REVIEW

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/GenericEntity.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    public void setValueXml(String valueXml) {
+        this.valueXml = valueXml;
+    }
+
+    @JsonIgnore
+    @Override
+    public AnnotatedEntity<GatewayEntity> getAnnotatedEntity() {
+        return createAnnotatedEntity();
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/GenericEntity.java:62]: this creates AnnotatedEntity instance every time

Comment (uppoju) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/GenericEntity.java:62]: Annotations are configured at design time so we can store these values in instance variable.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/GenericEntityBuilder.java [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+        Element genericEntityElement = createElementWithAttributesAndChildren(
+                document,
+                GENERIC_ENTITY,
+                ImmutableMap.of(ATTRIBUTE_ID, id),
+                createElementWithTextContent(document, NAME, name),
+                createElementWithTextContent(document, DESCRIPTION, genericEntity.getDescription()),
+                createElementWithTextContent(document, ENTITY_CLASS_NAME, genericEntity.getEntityClassName()),
+                // Enabled element must be after Description and EntityClassName
+                createElementWithTextContent(document, ENABLED, Boolean.TRUE.toString()),
+                createElementWithTextContent(document, VALUE_XML, genericEntity.getValueXml())
```
Comment (uppoju) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/GenericEntityBuilder.java:72]: We have to remove id from xml in GenericEntityLinker and update the xml in this builder

<l7:ValueXml>&lt;?xml version="1.0" encoding="UTF-8"?&gt;
&lt;java version="1.8.0_242" class="java.beans.XMLDecoder"&gt;
 &lt;object class="com.l7tech.external.assertions.http2.routing.model.Http2ClientConfigurationEntity"&gt;
  &lt;void property="description"&gt;
   &lt;string&gt;&lt;/string&gt;
  &lt;/void&gt;
  &lt;void property="id"&gt;
   &lt;string&gt;b9020f62133b1e4d09c087abcbca2666&lt;/string&gt;
  &lt;/void&gt;
  &lt;void property="name"&gt;
   &lt;string&gt;http2config&lt;/string&gt;
  &lt;/void&gt;
  &lt;void property="valueXml"&gt;
   &lt;string&gt;&lt;/string&gt;
  &lt;/void&gt;
  &lt;void property="version"&gt;
   &lt;int&gt;1&lt;/int&gt;
  &lt;/void&gt;
 &lt;/object&gt;
&lt;/java&gt;
</l7:ValueXml>

Comment (nikhilahuja) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/GenericEntityBuilder.java:72]: @uppoju We are not modifying the ValueXml and I don't think this Id has any effect on the object that gets created so not required I suppose. I tested the deployment with the same ID as in export and by modifying the ID in `@bundle-entity` annotation. Both worked!

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/GenericEntity.java (hunk 1 above)
Comment (nikhilahuja) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/beans/GenericEntity.java]: Fixed!

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/GenericEntityBuilder.java (hunk 2 above)
Comment (uppoju) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/GenericEntityBuilder.java:67]: OK

**Pull request merged**